commit id:

In [85]:
# Imports and pip installations (if needed)
import pandas as pd              

from sklearn.model_selection import train_test_split    
from sklearn.linear_model import LogisticRegression     
from sklearn.svm import SVC                        
from sklearn.neural_network import MLPClassifier 
from sklearn.neighbors import KNeighborsClassifier  


# Part 1: Load the dataset

In [86]:
# Load the dataset (load remotely or locally)

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
data = pd.read_csv("iris.csv")
print(data.head(15))
print(data.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

# Part 2: Split the dataset into train and test

In [87]:
# Take the dataset and split it into our features (X) and label (y)
x = data[["sepal_length", "sepal_width", "petal_length", "petal_width"]]
y = data["species"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=42)

# Part 3: Logistic Regression

In [88]:
# i. Use sklearn to train a LogisticRegression model on the training set
logistic_model = LogisticRegression(max_iter=200)
logistic_model.fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = pd.DataFrame({
    "sepal_length": [5], 
    "sepal_width": [3],
    "petal_length": [1],
    "petal_width": [0.2]
})

logistic_model_probabilities = logistic_model.predict_proba(sample)
logistic_model_prediction = logistic_model.predict(sample)[0]
labels = target_names = data["species"].unique()
print("Predicted Probabilities:")
for label, prob in zip(labels, logistic_model_probabilities[0]):
    print(f"Probability of {label}: {prob}")

print(f"Predicted class: {logistic_model_prediction}")
print()

# iii. Report on the score for Logistic regression model, what does the score measure?
logistic_model_score = logistic_model.score(x_test, y_test)
print("Logistic Regression Score:", logistic_model_score)
print()

# iv. Extract the coefficents and intercepts for the boundary line(s)
logistic_model_coefficients = logistic_model.coef_
logistic_model_intercepts = logistic_model.intercept_

def equation(coeffs, intercept):
    return str(intercept) + " + " + str(coeffs[0]) + "x_1 + " + str(coeffs[1]) + "x_2 + " + str(coeffs[2]) + "x_3+ " + str(coeffs[3]) + "x_4" 

for i in range(3):
    print("f_" + str(i+1) + " = " + equation(logistic_model_coefficients[i], logistic_model_intercepts[i]))

Predicted Probabilities:
Probability of setosa: 0.9856675609964477
Probability of versicolor: 0.014332431667528359
Probability of virginica: 7.336024121542721e-09
Predicted class: setosa

Logistic Regression Score: 1.0

f_1 = 9.543634962591975 + -0.4283368821275909x_1 + 0.9673218918184181x_2 + -2.4474144705243437x_3+ -1.03779729472909x_4
f_2 = 1.8988147964788542 + 0.5130401581400084x_1 + -0.22029920805594355x_2 + -0.21606169206706283x_3+ -0.8445409669682873x_4
f_3 = -11.442449759071 + -0.08470327601242085x_1 + -0.7470226837624787x_2 + 2.663476162591394x_3+ 1.8823382616973798x_4


Predicted Probabilities:

Probability of setosa: 0.9856675609964477

Probability of versicolor: 0.014332431667528359

Probability of virginica: 7.336024121542721e-09

Predicted class: setosa

Logistic Regression Score: 1.0

Meaning of Score: The Logistic Regression Score represents how accurate the predictions of the test data were. A score of 1.0 means that the model made perfect predictions for every data point in the test set.

$f_1 = 9.543634962591975 + -0.4283368821275909x_1 + 0.9673218918184181x_2 + -2.4474144705243437x_3+ -1.03779729472909x_4$

$f_2 = 1.8988147964788542 + 0.5130401581400084x_1 + -0.22029920805594355x_2 + -0.21606169206706283x_3+ -0.8445409669682873x_4$

$f_3 = -11.442449759071 + -0.08470327601242085x_1 + -0.7470226837624787x_2 + 2.663476162591394x_3+ 1.8823382616973798x_4$

# Part 4: Support Vector Machine

In [89]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm = SVC(probability=True)
svm.fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
svm_probabilities = svm.predict_proba(sample)
svm_prediction = svm.predict(sample)[0]
print("Predicted Probabilities:")
for label, prob in zip(labels, svm_probabilities[0]):
    print(f"Probability of {label}: {prob}")

print(f"Predicted class: {svm_prediction}")
print()

# iii. Report on the score for the SVM, what does the score measure?
svm_score = svm.score(x_test, y_test)
print("SVM Score:", svm_score)

Predicted Probabilities:
Probability of setosa: 0.9806347102698405
Probability of versicolor: 0.011113117171488503
Probability of virginica: 0.00825217255867099
Predicted class: setosa

SVM Score: 1.0


Predicted Probabilities:

Probability of setosa: 0.9806347102698405

Probability of versicolor: 0.011113117171488503

Probability of virginica: 0.00825217255867099

Predicted class: setosa

SVM Score: 1.0

Meaning of Score: The Support Vector Machine Score represents how accurate the predictions of the test data were. A score of 1.0 means that the model made perfect predictions for every data point in the test set. 

# Part 5: Neural Network

In [90]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(hidden_layer_sizes=(10,), max_iter=5000, random_state=42)
mlp.fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
mlp_probabilities = mlp.predict_proba(sample)
mlp_prediction = mlp.predict(sample)[0]
print("Predicted Probabilities:")
for label, prob in zip(labels, mlp_probabilities[0]):
    print(f"Probability of {label}: {prob}")

print(f"Predicted class: {mlp_prediction}")
print()

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp.score(x_test, y_test)
print("Neural Network Score:", mlp_score)
print()

# iv: Experiment with different options for the neural network, report on your best configuration
mlp_2 = MLPClassifier(hidden_layer_sizes=(20,), max_iter=10000, random_state=42)
mlp_2.fit(x_train, y_train)

mlp_2_probabilities = mlp_2.predict_proba(sample)
mlp_2_prediction = mlp_2.predict(sample)[0]
print("Predicted Probabilities:")
for label, prob in zip(labels, mlp_2_probabilities[0]):
    print(f"Probability of {label}: {prob}")

print(f"Predicted class: {mlp_2_prediction}")
print()

mlp_2_score = mlp_2.score(x_test, y_test)
print("Alternative Neural Network Score:", mlp_2_score)

Predicted Probabilities:
Probability of setosa: 0.988842674933009
Probability of versicolor: 0.011151976135902497
Probability of virginica: 5.348931088479587e-06
Predicted class: setosa

Neural Network Score: 0.9333333333333333

Predicted Probabilities:
Probability of setosa: 0.9939374917614512
Probability of versicolor: 0.006062508220451671
Probability of virginica: 1.8097127891456986e-11
Predicted class: setosa

Alternative Neural Network Score: 1.0


Predicted Probabilities:

Probability of setosa: 0.988842674933009

Probability of versicolor: 0.011151976135902497

Probability of virginica: 5.348931088479587e-06

Predicted class: setosa

Neural Network Score: 0.9333333333333333

Meaning of Score: The Neural Network Score represents how accurate the predictions of the test data were. A score of 0.93 means that the model made correct predictions appropriately 93% of the time.

Predicted Probabilities:

Probability of setosa: 0.9939374917614512

Probability of versicolor: 0.006062508220451671

Probability of virginica: 1.8097127891456986e-11

Predicted class: setosa

Alternative Neural Network Score: 1.0

Meaning of Score: The Neural Network Score represents how accurate the predictions of the test data were. A score of 1.0 means that the model made perfect predictions for every data point in the test set.

The best configurations use more hidden layers and more maximum iterations.

# Part 6: K-Nearest Neighbors

In [91]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
knn_probabilities = knn.predict_proba(sample)
knn_prediction = knn.predict(sample)[0]
print("Predicted Probabilities:")
for label, prob in zip(labels, knn_probabilities[0]):
    print(f"Probability of {label}: {prob}")

print(f"Predicted class: {knn_prediction}")
print()

# iii. Report on the score for kNN, what does the score measure?
knn_score = knn.score(x_test, y_test)
print("kNN Score:", knn_score)

Predicted Probabilities:
Probability of setosa: 1.0
Probability of versicolor: 0.0
Probability of virginica: 0.0
Predicted class: setosa

kNN Score: 1.0


Predicted Probabilities:

Probability of setosa: 1.0

Probability of versicolor: 0.0

Probability of virginica: 0.0

Predicted class: setosa

Meaning of Score: The K-Nearest Neighbors Score represents how accurate the predictions of the test data were. A score of 1.0 means that the model made perfect predictions for every data point in the test set.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

This notebook discovers the difference between certain classification models. Every model except for the first neural network exactly predicted the classes of each data point in the test set 100% of the time. I think that this is because the data points for each class in the dataset are really close together and far from other classes, and because the neural network did not have enough connections between hidden layers to find the connections between the features and labels. It also surprised me how simple it is to train such models using python libraries.  